contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -4627,4 +4635,93 @@ static Node* optimize_explicit_cast(ParseState *pstate, Node *node)
 		}
 	}
 	return node;
-}
\ No newline at end of file
+}
+
+/* 
+ * Portals caches the entire or remaining result set on transaction 
+ * boundaries. Transaction commands will fail if this portal fetch
+ * fails. To handle this we silently fail and close SCROLL cursors
+ * although for NO_SCROLL cursors we make do with whatever is fetched.
+ * We also need to restore to the correct interrupt count value as we
+ * could already be inside a interrupt holdoff section. Also we do not
+ * rethrow the error here and end up in a dirty state, so some 
+ * transaction clean up should be done after this.
+ */
+static void
+pltsql_persist_holdable_cursor_executor(Portal portal, QueryDesc *queryDesc,
+										ScanDirection direction)
+{
+	uint32 saveInterruptHoldoffCount = InterruptHoldoffCount;
+	uint32 saveQueryCancelHoldoffCount = QueryCancelHoldoffCount;
+	int	save_pltsql_non_tsql_proc_entry_count = pltsql_non_tsql_proc_entry_count;
+	int	save_pltsql_sys_func_entry_count= pltsql_sys_func_entry_count;
+	
+	/* do this for pltsql open cursors only */
+	if (portal->name != NULL && pltsql_is_cursor_open(portal->name))
+	{
+		int          	prev_spi_stack_depth = SPI_get_depth();
+		MemoryContext 	old_context = CurrentMemoryContext;
+		
+		PG_TRY();
+		{
+			ExecutorRun(queryDesc, direction, 0L, false);
+		}
+		PG_CATCH();
+		{
+			MemoryContext ectx;
+			ErrorData   	*error_data;
+			CmdType		operation;
+			int     	current_spi_stack_depth = SPI_get_depth();
+			bool    	sendTuples;
+			int     	rc;
+			
+			/* implicitly close SCROLL type cursors on error */
+			if (portal->cursorOptions & CURSOR_OPT_SCROLL)
+			{
+				MarkPortalFailed(portal);
```

Comment:
1. If we were executing a transaction command, the command will fail since we aborted midway.
2. If we were already inside our error handling, throwing the error here will pull us out of babelfish error handling.